File: src/agents/load.ts
import {
  existsSync,
  readdirSync,
  readFileSync,
  realpathSync,
  statSync,
} from "node:fs";
import { homedir } from "node:os";
import { basename, join } from "node:path";
import { getOriginalCwd } from "../workspace.js";
import { parseFrontmatter } from "../skills.js";
import type { AgentDefinition } from "./types.js";

/**
 * A user-defined agent loaded from a `.deepdive/agents/*.md` file. Mirrors
 * Claude Code's `.claude/agents/*.md` subagents: YAML-ish frontmatter declares
 * the agent's identity and tool scope, and the markdown body becomes the
 * persona system prompt. Carries provenance for the `/agents` listing.
 */
export interface LoadedAgent extends AgentDefinition {
  source: "user" | "project";
  filePath: string;
}

function agentDirs(): Array<{ dir: string; source: LoadedAgent["source"] }> {
  // User first, project second — project wins on an agentType clash (the
  // dedup in the registry keeps the LAST entry for a given type).
  return [
    { dir: join(homedir(), ".deepdive", "agents"), source: "user" },
    { dir: join(getOriginalCwd(), ".deepdive", "agents"), source: "project" },
  ];
}

/** The raw text between the leading `---` markers (the frontmatter region). */
function frontmatterBlock(raw: string): string {
  if (!raw.startsWith("---")) return "";
  const end = raw.indexOf("\n---", 3);
  if (end === -1) return "";
  return raw.slice(3, end);
}

function unquote(s: string): string {
  const v = s.trim();
  if (
    (v.startsWith('"') && v.endsWith('"')) ||
    (v.startsWith("'") && v.endsWith("'"))
  ) {
    return v.slice(1, -1).trim();
  }
  return v;
}

/**
 * Parse a frontmatter tool-list field into an AgentDefinition allowlist,
 * supporting every form the emulated Claude Code `.claude/agents/*.md` files
 * use (the shared single-line parseFrontmatter can't see YAML block lists):
 *
 *   tools: read_file, grep        → ["read_file", "grep"]   (comma)
 *   tools: [read_file, grep]      → ["read_file", "grep"]   (inline array)
 *   tools:\n  - read_file\n  - grep → ["read_file", "grep"] (block list)
 *   tools: *  /  tools: all        → undefined (inherit ALL)
 *   tools: none                    → []        (no tools)
 *   tools:  (empty, no list items) → undefined (inherit ALL — never a silent
 *                                     zero-tool agent)
 *   key absent                     → undefined (inherit ALL)
 */
export function parseToolList(block: string, key: string): string[] | undefined {
  const lines = block.split(/\r?\n/);
  const keyRe = new RegExp(`^${key}:\\s*(.*)$`);
  for (let i = 0; i < lines.length; i++) {
    const m = keyRe.exec(lines[i]!);
    if (!m) continue;

    const inline = unquote(m[1]!);
    if (inline) {
      const lower = inline.toLowerCase();
      if (lower === "*" || lower === "all") return undefined;
      if (lower === "none") return [];
      const items = inline
        .replace(/^\[/, "")
        .replace(/\]$/, "")
        .split(",")
        .map((t) => unquote(t))
        .filter(Boolean);
      return items.length ? items : undefined;
    }

    // Empty value → collect a following YAML block list (`  - item`).
    const items: string[] = [];
    for (let j = i + 1; j < lines.length; j++) {
      const line = lines[j]!;
      const item = /^\s*-\s+(.*)$/.exec(line);
      if (item) {
        const value = unquote(item[1]!);
        if (value) items.push(value);
      } else if (line.trim() === "" || /^\s+\S/.test(line)) {
        // blank line or indented non-list line — part of the block, skip
        continue;
      } else {
        break; // next top-level key
      }
    }
    // Empty `tools:` with no list items inherits ALL, not zero tools.
    return items.length ? items : undefined;
  }
  return undefined;
}

function firstContentLine(content: string, fallback: string): string {
  for (const line of content.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#") || trimmed.startsWith("```")) {
      continue;
    }
    return trimmed.replace(/^[-*]\s+/, "");
  }
  return fallback;
}

function readAgentsDir(
  dir: string,
  source: LoadedAgent["source"],
): LoadedAgent[] {
  if (!existsSync(dir)) return [];
  let entries: string[];
  try {
    entries = readdirSync(dir);
  } catch {
    return [];
  }

  const agents: LoadedAgent[] = [];
  for (const entry of entries) {
    if (!entry.endsWith(".md")) continue;
    const filePath = join(dir, entry);
    try {
      if (!statSync(filePath).isFile()) continue;
      const raw = readFileSync(filePath, "utf-8");
      const { frontmatter, content } = parseFrontmatter(raw);
      const block = frontmatterBlock(raw);
      // `name` identifies the agent (what the model passes as subagent_type);
      // fall back to the filename so a bare `reviewer.md` still works.
      const agentType = (frontmatter.name || basename(entry, ".md")).trim();
      if (!agentType) continue;
      const persona = content.trim();
      const whenToUse =
        frontmatter.description || firstContentLine(content, agentType);
      agents.push({
        agentType,
        whenToUse,
        tools: parseToolList(block, "tools"),
        disallowedTools: parseToolList(block, "disallowedTools"),
        model: frontmatter.model || undefined,
        getSystemPrompt: () => persona,
        source,
        filePath,
      });
    } catch {
      // Ignore a broken/unreadable agent file; the rest must still load.
    }
  }
  return agents;
}

/**
 * Load every user-defined agent from the user and project `.deepdive/agents`
 * directories. Project agents are returned AFTER user agents so a same-named
 * project agent overrides the user one when the registry dedups by agentType
 * (last-wins). A file resolving to the same realpath as one already seen is
 * skipped (handles a symlinked/duplicated dir).
 */
export function loadCustomAgents(): LoadedAgent[] {
  const seen = new Set<string>();
  const agents: LoadedAgent[] = [];
  for (const { dir, source } of agentDirs()) {
    for (const agent of readAgentsDir(dir, source)) {
      let fileId: string;
      try {
        fileId = realpathSync(agent.filePath);
      } catch {
        fileId = agent.filePath;
      }
      if (seen.has(fileId)) continue;
      seen.add(fileId);
      agents.push(agent);
    }
  }
  return agents;
}

File: deepdive-rs/crates/deepdive-core/src/agents/load.rs
//! Load user/project agents from `.deepdive/agents/*.md`. Port of
//! `src/agents/load.ts`.

use super::types::AgentDefinition;
use crate::skills::{parse_frontmatter, AssetSource};
use crate::workspace::original_cwd;
use regex::Regex;
use std::path::{Path, PathBuf};
use std::sync::LazyLock;

static BLOCK_ITEM_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^\s*-\s+(.*)$").unwrap());
static INDENTED_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^\s+\S").unwrap());
static LIST_BULLET_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^[-*]\s+").unwrap());

#[derive(Debug, Clone)]
pub struct LoadedAgent {
    pub def: AgentDefinition,
    pub source: AssetSource,
    pub file_path: String,
}

fn unquote(value: &str) -> String {
    let v = value.trim();
    if v.len() >= 2
        && ((v.starts_with('"') && v.ends_with('"')) || (v.starts_with('\'') && v.ends_with('\'')))
    {
        v[1..v.len() - 1].trim().to_string()
    } else {
        v.to_string()
    }
}

fn agent_dirs() -> Vec<(PathBuf, AssetSource)> {
    let home = std::env::var_os("HOME")
        .or_else(|| std::env::var_os("USERPROFILE"))
        .map(PathBuf::from)
        .unwrap_or_default();
    vec![
        (home.join(".deepdive").join("agents"), AssetSource::User),
        (
            original_cwd().join(".deepdive").join("agents"),
            AssetSource::Project,
        ),
    ]
}

fn frontmatter_block(raw: &str) -> String {
    if !raw.starts_with("---") {
        return String::new();
    }
    match raw[3..].find("\n---") {
        Some(rel) => raw[3..3 + rel].to_string(),
        None => String::new(),
    }
}

/// Parse a frontmatter tool-list field (comma / inline-array / block-list /
/// `*`/`all`/`none`) into an allowlist. `None` = inherit all.
pub fn parse_tool_list(block: &str, key: &str) -> Option<Vec<String>> {
    let key_re = Regex::new(&format!(r"^{}:\s*(.*)$", regex::escape(key))).unwrap();
    let lines: Vec<&str> = block
        .split('\n')
        .map(|l| l.strip_suffix('\r').unwrap_or(l))
        .collect();

    for (i, line) in lines.iter().enumerate() {
        let Some(caps) = key_re.captures(line) else {
            continue;
        };
        let inline = unquote(&caps[1]);
        if !inline.is_empty() {
            let lower = inline.to_lowercase();
            if lower == "*" || lower == "all" {
                return None;
            }
            if lower == "none" {
                return Some(Vec::new());
            }
            // strip one leading `[` then one trailing `]` (each optional).
            let s1 = inline.strip_prefix('[').unwrap_or(&inline);
            let stripped = s1.strip_suffix(']').unwrap_or(s1);
            let items: Vec<String> = stripped
                .split(',')
                .map(unquote)
                .filter(|t| !t.is_empty())
                .collect();
            return if items.is_empty() { None } else { Some(items) };
        }

        // empty value → collect a following YAML block list (`  - item`).
        let mut items = Vec::new();
        for line2 in &lines[i + 1..] {
            if let Some(item) = BLOCK_ITEM_RE.captures(line2) {
                let value = unquote(&item[1]);
                if !value.is_empty() {
                    items.push(value);
                }
            } else if line2.trim().is_empty() || INDENTED_RE.is_match(line2) {
                continue;
            } else {
                break;
            }
        }
        return if items.is_empty() { None } else { Some(items) };
    }
    None
}

fn first_content_line(content: &str, fallback: &str) -> String {
    for line in content.split('\n') {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') || trimmed.starts_with("```") {
            continue;
        }
        return LIST_BULLET_RE.replace(trimmed, "").to_string();
    }
    fallback.to_string()
}

fn read_agents_dir(dir: &Path, source: AssetSource) -> Vec<LoadedAgent> {
    let Ok(entries) = std::fs::read_dir(dir) else {
        return Vec::new();
    };
    let mut agents = Vec::new();
    for entry in entries.flatten() {
        let file_path = entry.path();
        if file_path.extension().and_then(|e| e.to_str()) != Some("md") {
            continue;
        }
        if !file_path.is_file() {
            continue;
        }
        let Ok(raw) = std::fs::read_to_string(&file_path) else {
            continue;
        };
        let (frontmatter, content) = parse_frontmatter(&raw);
        let block = frontmatter_block(&raw);
        let file_stem = file_path
            .file_stem()
            .map(|s| s.to_string_lossy().into_owned())
            .unwrap_or_default();
        let agent_type = frontmatter
            .get("name")
            .filter(|s| !s.is_empty())
            .cloned()
            .unwrap_or(file_stem)
            .trim()
            .to_string();
        if agent_type.is_empty() {
            continue;
        }
        let persona = content.trim().to_string();
        let when_to_use = frontmatter
            .get("description")
            .filter(|s| !s.is_empty())
            .cloned()
            .unwrap_or_else(|| first_content_line(&content, &agent_type));
        agents.push(LoadedAgent {
            def: AgentDefinition {
                agent_type,
                when_to_use,
                tools: parse_tool_list(&block, "tools"),
                disallowed_tools: parse_tool_list(&block, "disallowedTools").unwrap_or_default(),
                model: frontmatter.get("model").filter(|s| !s.is_empty()).cloned(),
                system_prompt: persona,
            },
            source,
            file_path: file_path.to_string_lossy().into_owned(),
        });
    }
    agents
}

/// Load every user + project agent. Project after user so project wins on a
/// type clash (registry dedup is last-wins). Realpath dedup handles symlinks.
pub fn load_custom_agents() -> Vec<LoadedAgent> {
    let mut seen = std::collections::HashSet::new();
    let mut agents = Vec::new();
    for (dir, source) in agent_dirs() {
        for agent in read_agents_dir(&dir, source) {
            let file_id = std::fs::canonicalize(&agent.file_path)
                .map(|p| p.to_string_lossy().into_owned())
                .unwrap_or_else(|_| agent.file_path.clone());
            if seen.insert(file_id) {
                agents.push(agent);
            }
        }
    }
    agents
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn tool_list_forms() {
        assert_eq!(
            parse_tool_list("tools: read_file, grep", "tools"),
            Some(vec!["read_file".to_string(), "grep".to_string()])
        );
        assert_eq!(
            parse_tool_list("tools: [read_file, grep]", "tools"),
            Some(vec!["read_file".to_string(), "grep".to_string()])
        );
        assert_eq!(
            parse_tool_list("tools:\n  - read_file\n  - grep", "tools"),
            Some(vec!["read_file".to_string(), "grep".to_string()])
        );
        assert_eq!(parse_tool_list("tools: *", "tools"), None);
        assert_eq!(parse_tool_list("tools: all", "tools"), None);
        assert_eq!(parse_tool_list("tools: none", "tools"), Some(vec![]));
        assert_eq!(parse_tool_list("tools:", "tools"), None); // empty → inherit all
        assert_eq!(parse_tool_list("other: x", "tools"), None); // key absent
    }

    #[test]
    fn block_list_stops_at_next_key() {
        let block = "tools:\n  - read_file\n  - grep\nmodel: fast";
        assert_eq!(
            parse_tool_list(block, "tools"),
            Some(vec!["read_file".to_string(), "grep".to_string()])
        );
    }
}
